proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1584 @@
+
+```yaml
+#@data/values-schema
+---
+dex:
+  #@schema/validation min_len=1
+  namespace: ""
+  #@schema/validation min_len=1
+  username: ""
+```
+
+This case is so common (it is our understanding that most string-based data values are meant to be non-empty), we offer this convenience:
+
+```yaml
+#@data/values-schema
+#@schema/validation-defaults-for-strings min_len=1
+---
+dex:
+  namespace: ""
+  username: ""
+```
+
+Which is equivalent to the prior example, but applies the default to any descendent that contains a string-typed value. If a particular node already has an `@assert/validate` annotation, it is skipped.
+
+**Array Values**
+
+Similarly, a non-empty value of an array has at least one element:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  oauth2:
+    #@schema/validation min_len=1
+    responseTypes:
+    - ""
+```
+
+Array-typed values are less common. And when present are not necessarily _as_ common to require a non-empty value as string-typed values. So, there is no corresponding shorthand annotation for requiring a non-empty array.
+
+
+##### Required input for types with empty value
+
+For some integer-typed (or float-typed) data values, a zero value may be invalid and thus can represent "empty".
+
+```yaml
+#@data/values-schema
+dex:
+  #@schema/validation min=1024, max=65535
+  port: 0
+```
+
+Authors are encouraged to favor this approach over [marking the data values as nullable](#required-input-for-types-with-no-empty-value) to keep their schema as simple as possible.
+ 
+##### Required input for types with no empty value
+
+Some types have no natural (or easily detectable) "empty" value. In these cases, authors can force the consumer to supply a value by marking the value as "nullable" and constraint it to be "not null":
+
+```yaml
+#@data/values-schema
+---
+#@schema/nullable
+#@schema/validation not_null=("Cloud credentials are required.", True)
+credential:
+  name:
+  secretContents:
+    cloud:
+```
+
+_(also considered: [`@schema/required`](#schemarequired))._
+
+#### Use Case: Enumeration
+
+Values can be constrained to be one of a finite site (i.e. of an enumeration).
+
+```yaml
+#@data/values-schema
+---
+volumeSnapshotLocation:
+  spec:
+    #@schema/validation one_of=["aws", "azure", "vsphere"]
+    provider: ""
+```
+
+#### Use Case: Union Structures
+
+An author can require that exactly one of a set of sibling node has a value:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  #@schema/validation one_not_null=["oidc", "ldap"]
+  config:
+    #@schema/nullable
+    oidc:
+      CLIENT_ID: null #! required if oidc enabled
+      CLIENT_SECRET: null #! required if oidc enabled
+      issuer: null #! <OIDC_IDP_URL> is required if oidc enabled
+    #@schema/nullable
+    ldap:
+      host: null #! <LDAP_HOST> is required if ldap enabed
+      bindDN: null
+      bindPW: null
+```
+
+This approach is simpler than (and therefore recommended over) using a discriminator:
```

Comment:
Yeah, most of the "values.yaml" that come across my desk use the discriminator approach; that's a good point.
I believe that's true, in part, because it's kinda the only way it can be done in most templating systems.

Providing proper validation (you're hinting at a potential sharp-edge, here) gets tricky with a discriminator. For example, for a section that is _not_ selected, either all the default values need to be valid, or there needs to be a way to disable those validations, or validations must be hoisted to their parent node.

We've got some work ahead of us to make sure all this is clear so users can make an informed decision _and_ not get lost/mired in a mess if they _must_ use the discriminator approach.

The team is still open to ideas that could make the discriminator approach work so long as it doesn't result in undue complexity in the tool overall. Things like being able to specify paths or reference nodes above the annotated node tend to lead to accidental complexity / rigidity.

This is a great topic to dig into; thank you for bringing this up!!

Reply:
Captured this as a note for docs implementation: https://github.com/vmware-tanzu/carvel/pull/331/files#diff-a81faf6d33e82e22eee1083ee484a23eda8a928b046116a2cb516e58086984b8R1585